README
NAME
    MooseX::XSAccessor - use Class::XSAccessor to speed up Moose accessors

SYNOPSIS
       package MyClass;
   
       use Moose;
       use MooseX::XSAccessor;
   
       has foo => (...);

DESCRIPTION
    This module accelerates Moose-generated accessor, reader, writer and
    predicate methods using Class::XSAccessor. You get a speed-up for no extra
    effort. It is automatically applied to every attribute in the class.

    The use of the following features of Moose attributes prevents a reader
    from being accelerated:

    *   Lazy builder or lazy default.

    *   Auto-deref. (Does anybody use this anyway??)

    The use of the following features prevents a writer from being
    accelerated:

    *   Type constraints (except `Any`; `Any` is effectively a no-op).

    *   Triggers

    *   Weak references

    An `rw` accessor is effectively a reader and a writer glued together, so
    both of the above lists apply.

    Predicates can always be accelerated, provided you're using
    Class::XSAccessor 1.17 or above.

    Clearers can not be accelerated (as of current versions of
    Class::XSAccessor).

  Functions
    This module also provides one function, which is not exported so needs to
    be called by its full name.

    `MooseX::XSAccessor::is_xs($sub)`
        Returns a boolean indicating whether a sub is an XSUB.

        $sub may be a coderef, Class::MOP::Method object, or a qualified sub
        name as a string (e.g. "MyClass::foo").

  Chained accessors and writers
    MooseX::XSAccessor can detect chained accessors and writers created using
    MooseX::Attribute::Chained, and can accelerate those too.

       package Local::Class;
       use Moose;
       use MooseX::XSAccessor;
       use MooseX::Attribute::Chained;
   
       has foo => (traits => ["Chained"], is => "rw");
       has bar => (traits => ["Chained"], is => "ro", writer => "_set_bar");
       has baz => (                       is => "rw");  # not chained
   
       my $obj = "Local::Class"->new;
       $obj->foo(1)->_set_bar(2);
       print $obj->dump;

  Lvalue accessors
    MooseX::XSAccessor will detect lvalue accessors created with
    MooseX::LvalueAttribute and, by default, skip accelerating them.

    However, by setting $MooseX::XSAccessor::LVALUE to true (preferably using
    the `local` Perl keyword), you can force it to accelerate those too. This
    introduces a visible change in behaviour though. MooseX::LvalueAttribute
    accessors normally allow two patterns for setting the value:

       $obj->foo = 42;   # as an lvalue
       $obj->foo(42);    # as a method call

    However, once accelerated, they may *only* be set as an lvalue. For this
    reason, setting $MooseX::XSAccessor::LVALUE to true is considered an
    experimental feature.

HINTS
    *   Make attributes read-only when possible. This means that type
        constraints and coercions will only apply to the constructor, not the
        accessors, enabling the accessors to be accelerated.

    *   If you do need a read-write attribute, consider making the main
        accessor read-only, and having a separate writer method. (Like
        MooseX::SemiAffordanceAccessor.)

    *   Make defaults eager instead of lazy when possible, allowing your
        readers to be accelerated.

    *   If you need to accelerate just a specific attribute, apply the
        attribute trait directly:

           package MyClass;
   
           use Moose;
   
           has foo => (
              traits => ["MooseX::XSAccessor::Trait::Attribute"],
              ...,
           );

    *   If you don't want to add a dependency on MooseX::XSAccessor, but do
        want to use it if it's available, the following code will use it
        optionally:

           package MyClass;
   
           use Moose;
           BEGIN { eval "use MooseX::XSAccessor" };
   
           has foo => (...);

CAVEATS
    *   Calling a writer method without a parameter in Moose does not raise an
        exception:

           $person->set_name();    # sets name attribute to "undef"

        However, this is a fatal error in Class::XSAccessor.

    *   MooseX::XSAccessor does not play nice with attribute traits that alter
        accessor behaviour, or define additional accessors for attributes.
        MooseX::SetOnce is an example thereof. MooseX::Attribute::Chained is
        handled as a special case.

    *   MooseX::XSAccessor only works on blessed hash storage; not e.g.
        MooseX::ArrayRef or MooseX::InsideOut. MooseX::XSAccessor is usually
        able to detect such situations and silently switch itself off.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=MooseX-XSAccessor>.

SEE ALSO
    MooseX::XSAccessor::Trait::Attribute.

    Moose, Moo, Class::XSAccessor.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2013 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
MooseX-XSAccessor
=================

Created:      2013-06-13
Home page:    <https://metacpan.org/release/MooseX-XSAccessor>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=MooseX-XSAccessor>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.007	2013-10-30

 [ Test Suite ]
 - Skip the new version of moose_bad_coerce.t if Moose version is too old.

0.006	2013-10-30

 [ Test Suite ]
 - Avoid triggering silly deprecation warnings from
   MooseX::Attribute::Chained.
 - Pull latest attribute tests from Moose; the old versions of these test
   cases broke with Moose 2.11xx.

0.005	2013-08-27

 - Added: Integration with MooseX::LvalueAttribute.

0.004	2013-06-17

 - Added: Integration with MooseX::Attribute::Chained.
   Dagfinn Ilmari Mannsåker++
 - Improved `is_xs` implementation.
   Florian Ragwitz++

0.003	2013-06-17

 [ Documentation ]
 - Document the circumstances under which predicates and clearers can be
   acceleated (respectively: if Class::XSAccessor is new enough, and
   never).

 [ Packaging ]
 - Since 0.002 we shouldn't need to skip test `t/moose_default_undef.t`
   anymore.

0.002	2013-06-16

 [ Packaging ]
 - Support slightly older versions of Class::XSAccessor (back to 1.09).

 [ Other ]
 - Don't use Class::XSAccessor 1.16 and below to generate predicate
   methods, because their behaviour differs observably from Moose. If you
   want XS predicates, you'll need Class::XSAccessor 1.17 (which is not on
   CPAN yet).
   <http://purl.org/NET/cpan-uri/rt/ticket/86127>
 - Refactor overridden install_accessors into an 'after' method modifier,
   with the hope that this makes MooseX::XSAccessor play nicer with other
   MooseX modules.

0.001	2013-06-14	Stable release

0.000_02	2013-06-14

 [ Bug Fixes ]
 - Work around some edge cases.

 [ Documentation ]
 - Better documentation.

 [ Packaging ]
 - Better test cases.

 [ Other ]
 - Added: MooseX::XSAccessor::is_xs function.

0.000_01	2013-06-14	Developer release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "use Class::XSAccessor to speed up Moose accessors",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.013, CPAN::Meta::Converter version 2.120921",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "MooseX-XSAccessor",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        runtime => {
                          requires => { "Class::XSAccessor" => 1.09, "Moose" => "2.0600" },
                          suggests => { "MooseX::Attribute::Chained" => 0, "MooseX::LvalueAttribute" => 0 },
                        },
                        test => {
                          recommends => { "MooseX::Attribute::Chained" => 0, "MooseX::LvalueAttribute" => 0 },
                          requires   => {
                                          "Import::Into"   => "1.001000",
                                          "Test::Fatal"    => 0,
                                          "Test::Moose"    => 0,
                                          "Test::More"     => 0.96,
                                          "Test::Requires" => 0,
                                        },
                        },
                      },
  "provides"       => {
                        "MooseX::XSAccessor" => { file => "lib/MooseX/XSAccessor.pm", version => 0.007 },
                        "MooseX::XSAccessor::Trait::Attribute" => {
                          file => "lib/MooseX/XSAccessor/Trait/Attribute.pm",
                          version => 0.007,
                        },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => {
                                          web => "http://rt.cpan.org/Dist/Display.html?Queue=MooseX-XSAccessor",
                                        },
                        homepage     => "https://metacpan.org/release/MooseX-XSAccessor",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          web  => "https://github.com/tobyink/p5-moosex-xsaccessor",
                                        },
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/MooseX-XSAccessor/project",
                      },
  "version"        => 0.007,
  "x_contributors" => ["Florian Ragwitz (FLORA) <flora\@cpan.org>"],
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

